File: app/src/main/java/com/privat/contacts/data/cloud/model/AddressNet.java
package com.privat.contacts.data.cloud.model;

import androidx.annotation.NonNull;

import com.google.gson.annotations.SerializedName;

import org.jetbrains.annotations.NotNull;

public class AddressNet {
    @NotNull
    private final String city;
    @NotNull
    @SerializedName("street_name")
    private final String streetName;
    @NotNull
    @SerializedName("street_address")
    private final String streetAddress;
    @NotNull
    @SerializedName("zip_code")
    private final String zipCode;
    @NotNull
    @SerializedName("state")
    private final String state;
    @NotNull
    private final String country;
    @NotNull
    private final CoordinatesNet coordinates;
    public AddressNet(@NonNull String city,
                      @NonNull String streetName,
                      @NonNull String streetAddress,
                      @NonNull String zipCode,
                      @NonNull String state,
                      @NonNull String country,
                      @NonNull CoordinatesNet coordinates) {
        this.city = city;
        this.streetName = streetName;
        this.streetAddress = streetAddress;
        this.zipCode = zipCode;
        this.state = state;
        this.country = country;
        this.coordinates = coordinates;
    }
    public <T> T map(int userId, Mapper<T> mapper) {
        return mapper.map(
                userId,
                city,
                streetName,
                streetAddress,
                zipCode,
                state,
                country,
                coordinates);
    }
    public interface Mapper<T> {
        T map(int userId,
              @NonNull String city,
              @NonNull String streetName,
              @NonNull String streetAddress,
              @NonNull String zipCode,
              @NonNull String state,
              @NonNull String country,
              @NonNull CoordinatesNet coordinates);
    }
}

File: app/src/main/java/com/privat/contacts/data/cloud/model/AddressNet.kt
package com.privat.contacts.data.cloud.model

import com.google.gson.annotations.SerializedName

class AddressNet(private val city: String,
                 @field:SerializedName("street_name") private val streetName: String,
                 @field:SerializedName("street_address") private val streetAddress: String,
                 @field:SerializedName("zip_code") private val zipCode: String,
                 @field:SerializedName("state") private val state: String,
                 private val country: String,
                 private val coordinates: CoordinatesNet) {
    fun <T> map(userId: Int, mapper: Mapper<T>): T {
        return mapper.map(
                userId,
                city,
                streetName,
                streetAddress,
                zipCode,
                state,
                country,
                coordinates)
    }

    interface Mapper<T> {
        fun map(userId: Int,
                city: String,
                streetName: String,
                streetAddress: String,
                zipCode: String,
                state: String,
                country: String,
                coordinates: CoordinatesNet): T
    }
}